openwisp_monitoring/device/admin.py

```diff
@@ -126,6 +126,21 @@ def get_extra_context(self, pk=None):
             )
         return ctx
 
+    def metric_health(self, obj):
+        metric_rows = []
+        for metric in DeviceData.objects.get(pk=obj.pk).metrics.all():
```

Comment:
I think there's no need to generate another query to retrieve DeviceData, which is just a proxy model BTW and hence does not have particular DB content not available to device.

You should be able to just do:

`DeviceData(pk=obj.pk).metrics.all()`

This will be enough for the proxy model instance to know what query to make.

```python
for metric in DeviceData.objects.get(pk=obj.pk).metrics.all():
```